File: scripts/installer.js
#!/usr/bin/env node

/**
 * Node.js script to download and optionally install a VS Code extension from the Marketplace.
 * Usage: node installer.js [-i] <publisher.extension>
 */

const https = require('https');
const fs = require('fs');
const path = require('path');
const { execSync, spawnSync } = require('child_process');

function usage() {
  console.log('Usage: node installer.js [-i] <publisher.extension>');
  process.exit(1);
}

function requireCmd(cmds) {
  for (const cmd of cmds) {
    try {
      execSync(`command -v ${cmd}`, { stdio: 'ignore' });
    } catch {
      console.error(`Missing required command: ${cmd}`);
      process.exit(1);
    }
  }
}

function detectCodeCommand() {
  for (const cmd of ['code', 'codium', 'vscodium']) {
    try {
      execSync(`command -v ${cmd}`, { stdio: 'ignore' });
      console.log(`Using code CLI: ${cmd}`);
      return cmd;
    } catch {}
  }
  throw new Error("No VS Code CLI tool found (tried 'code', 'codium', 'vscodium').");
}

function fetchHtml(url) {
  return new Promise((resolve, reject) => {
    https.get(url, { headers: { 'User-Agent': 'Node.js' } }, (res) => {
      if (res.statusCode !== 200) {
        reject(`Failed to fetch Marketplace page (status: ${res.statusCode})`);
        return;
      }
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => resolve(data));
    }).on('error', reject);
  });
}

function extractAssetUri(html) {
  const assetUriMatch = html.match(/"assetUri"\s*:\s*"([^"]+)"/);
  const fallbackUriMatch = html.match(/"fallbackAssetUri"\s*:\s*"([^"]+)"/);
  if (assetUriMatch) {
    const assetUri = assetUriMatch[1];
    const vsixUrl = `${assetUri}/Microsoft.VisualStudio.Services.VSIXPackage`;
    console.log(`Found VSIX URL via assetUri: ${vsixUrl}`);
    return vsixUrl;
  } else if (fallbackUriMatch) {
    const fallbackUri = fallbackUriMatch[1];
    const vsixUrl = `${fallbackUri}/Microsoft.VisualStudio.Services.VSIXPackage`;
    console.log(`Found VSIX URL via fallbackAssetUri: ${vsixUrl}`);
    return vsixUrl;
  } else {
    throw new Error("Could not find assetUri or fallbackAssetUri in page.");
  }
}

function downloadVsix(url, filename) {
  return new Promise((resolve, reject) => {
    const file = fs.createWriteStream(filename);
    https.get(url, (res) => {
      if (res.statusCode !== 200) {
        reject('Failed to download the VSIX file.');
        return;
      }
      res.pipe(file);
      file.on('finish', () => {
        file.close(() => {
          console.log('Download complete.');
          resolve();
        });
      });
    }).on('error', (err) => {
      fs.unlinkSync(filename);
      reject(err.message);
    });
  });
}

function installExtension(codeCmd, filename) {
  console.log(`Installing extension using ${codeCmd} from ${filename}...`);
  const result = spawnSync(codeCmd, ['--install-extension', filename], { stdio: 'inherit' });
  if (result.status !== 0) {
    throw new Error('Installation failed.');
  }
  console.log('Installation complete!');
}

// ---- Main logic ----
const args = process.argv.slice(2);

if (args.length === 0) usage();

let installFlag = false;
let extensionId = '';

for (const arg of args) {
  if (arg === '-i') {
    installFlag = true;
  } else if (arg.includes('.')) {
    extensionId = arg;
  }
}
if (!extensionId) {
  console.error("Error: Missing extension name (e.g., publisher.extension)");
  usage();
}
if (!extensionId.includes('.')) {
  console.error("Extension ID must be in format 'publisher.extension'");
  process.exit(1);
}

const [publisher, extension] = extensionId.split('.');
const filename = `${extension}.vsix`;
const marketplaceUrl = `https://marketplace.visualstudio.com/items?itemName=${publisher}.${extension}&ssr=false#overview`;

(async () => {
  requireCmd(['curl']); // Used only for the check; node handles network
  console.log(`Fetching Marketplace page for ${publisher}.${extension}...`);
  let html;
  try {
    html = await fetchHtml(marketplaceUrl);
  } catch (err) {
    console.error('Error fetching Marketplace page:', err);
    process.exit(1);
  }

  let vsixUrl;
  try {
    vsixUrl = extractAssetUri(html);
  } catch (err) {
    console.error('Error extracting VSIX URL:', err);
    process.exit(1);
  }

  try {
    await downloadVsix(vsixUrl, filename);
  } catch (err) {
    console.error('Error downloading VSIX:', err);
    process.exit(1);
  }

  if (installFlag) {
    let codeCmd;
    try {
      codeCmd = detectCodeCommand();
    } catch (err) {
      console.error(err.message);
      process.exit(1);
    }
    try {
      installExtension(codeCmd, filename);
    } catch (err) {
      console.error(err.message);
      process.exit(1);
    }
  } else {
    console.log('Skipping installation. Use \'-i\' to install.');
  }
})();

File: scripts/installer.py
import re
import sys
import json
import shutil
import requests
import subprocess
from bs4 import BeautifulSoup


class VSCodeExtensionInstaller:
    MARKETPLACE_URL_1 = "https://marketplace.visualstudio.com/items?itemName={}&ssr=false#overview"
    MARKETPLACE_URL_2 = "https://marketplace.visualstudio.com/items?itemName={}&ssr=false#review-details"
    MARKETPLACE_URL_3 = "https://marketplace.visualstudio.com/items?itemName={}&ssr=false#qna"
    MARKETPLACE_URL_4 = "https://marketplace.visualstudio.com/items?itemName={}&ssr=false#version-history"

    def __init__(self, extension_id, install=False, editor_cmd=None):
        if "." not in extension_id:
            raise ValueError("Extension ID must be in format 'publisher.extension'")
        self.publisher, self.extension = extension_id.split(".")
        self.filename = f"{self.extension}.vsix"
        self.vsix_url = ""
        self.install_flag = install
        self.code_command = self.detect_code_command(editor_cmd)

    def detect_code_command(self, preferred=None):
        if preferred:
            if shutil.which(preferred):
                print(f"Using specified code CLI: {preferred}")
                return preferred
            else:
                raise EnvironmentError(f"Specified CLI '{preferred}' not found in PATH.")
        for cmd in ["code", "codium", "vscodium"]:
            if shutil.which(cmd):
                print(f"Using auto-detected code CLI: {cmd}")
                return cmd
        raise EnvironmentError("No VS Code CLI tool found (tried 'code', 'codium', 'vscodium').")

    def extract_asset_uri(self, html: str):
        asset_match = re.search(r'"assetUri"\s*:\s*"([^"]+)"', html)
        fallback_match = re.search(r'"fallbackAssetUri"\s*:\s*"([^"]+)"', html)

        if asset_match:
            asset_uri = asset_match.group(1)
            self.vsix_url = f"{asset_uri}/Microsoft.VisualStudio.Services.VSIXPackage"
            print(f"Found VSIX URL via assetUri: {self.vsix_url}")
        elif fallback_match:
            fallback_uri = fallback_match.group(1)
            self.vsix_url = f"{fallback_uri}/Microsoft.VisualStudio.Services.VSIXPackage"
            print(f"Found VSIX URL via fallbackAssetUri: {self.vsix_url}")
        else:
            print("Trying third fallback (BeautifulSoup JSON scan)...")
            soup = BeautifulSoup(html, "html.parser")
            scripts = soup.find_all("script", type="application/json")
            for script in scripts:
                try:
                    data = json.loads(script.string)
                    json_text = json.dumps(data)
                    match = re.search(
                        r'"assetType"\s*:\s*"Microsoft\.VisualStudio\.Services\.VSIXPackage"\s*,\s*"source"\s*:\s*"([^"]+)"',
                        json_text)
                    if match:
                        self.vsix_url = match.group(1)
                        print(f"Found VSIX URL via embedded JSON: {self.vsix_url}")
                        return
                except Exception:
                    continue
            raise Exception("Could not find assetUri, fallbackAssetUri, or embedded VSIXPackage source in page.")

    def fetch_download_url(self):
        print(f"Fetching Marketplace page for {self.publisher}.{self.extension}...")

        urls = [
            self.MARKETPLACE_URL_1.format(f"{self.publisher}.{self.extension}"),
            self.MARKETPLACE_URL_2.format(f"{self.publisher}.{self.extension}"),
            self.MARKETPLACE_URL_3.format(f"{self.publisher}.{self.extension}"),
            self.MARKETPLACE_URL_4.format(f"{self.publisher}.{self.extension}")
        ]

        for url in urls:
            print(f"Trying URL: {url}")
            response = requests.get(url)
            if response.status_code != 200:
                print(f"Failed to fetch: {url} (status: {response.status_code})")
                continue
            try:
                self.extract_asset_uri(response.text)
                return  # Success
            except Exception as e:
                print(f"Extraction failed on this URL: {e}")
                continue

        raise Exception("Failed to find VSIX download URL from all known Marketplace pages.")

    def download_vsix(self):
        print(f"Downloading VSIX to {self.filename}...")
        response = requests.get(self.vsix_url)
        if response.status_code != 200:
            raise Exception("Failed to download the VSIX file.")

        with open(self.filename, "wb") as f:
            f.write(response.content)
        print("Download complete.")

    def install_extension(self):
        if not self.install_flag:
            print("Skipping installation. Use '-i' to install.")
            return

        print(f"Installing extension using {self.code_command} from {self.filename}...")
        subprocess.run([self.code_command, "--install-extension", self.filename], check=True)
        print("Installation complete!")

    def run(self):
        self.fetch_download_url()
        self.download_vsix()
        self.install_extension()


if __name__ == "__main__":
    args = sys.argv[1:]

    if not args:
        print("Usage: python installer.py [-i] [-e editor] <publisher.extension>")
        sys.exit(1)

    install_flag = False
    extension_arg = None
    editor_cmd = None

    i = 0
    while i < len(args):
        if args[i] == "-i":
            install_flag = True
        elif args[i] == "-e":
            i += 1
            if i < len(args):
                editor_cmd = args[i]
            else:
                print("Error: -e flag requires an argument (code/codium/vscodium).")
                sys.exit(1)
        elif "." in args[i]:
            extension_arg = args[i]
        i += 1

    if not extension_arg:
        print("Error: Missing extension name (e.g., publisher.extension)")
        sys.exit(1)

    try:
        installer = VSCodeExtensionInstaller(extension_arg, install=install_flag, editor_cmd=editor_cmd)
        installer.run()
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)
